game-core/src/main/java/games/strategy/triplea/ui/unit/scroller/UnitScroller.java

```diff
@@ -0,0 +1,278 @@
+    centerPanel.add(selectUnitImagePanel);
+
+    final JButton nextUnit = new JButton(UnitScrollerIcon.RIGHT_ARROW.get());
+    nextUnit.setToolTipText(NEXT_UNITS_TOOLTIP);
+    nextUnit.setFocusable(false);
+    nextUnit.addActionListener(e -> centerOnNextMovableUnit());
+    centerPanel.add(nextUnit, BorderLayout.EAST);
+    centerPanel.add(Box.createHorizontalStrut(10));
+
+    panel.add(centerPanel);
+
+    final JButton skipButton = new JButton(UnitScrollerIcon.UNIT_SKIP.get());
+    skipButton.setToolTipText(SKIP_UNITS_TOOLTIP);
+    skipButton.setFocusable(false);
+    skipButton.setAlignmentX(JComponent.CENTER_ALIGNMENT);
+    skipButton.addActionListener(e -> skipCurrentUnits());
+    panel.add(skipButton, BorderLayout.SOUTH);
+    panel.add(Box.createVerticalStrut(3));
+
+    panel.add(SwingComponents.leftBox(movesLeftLabel));
+
+    panel.add(Box.createVerticalStrut(5));
+    return panel;
+  }
+
+  /** Centers the map on the current territory shown in the unit scroller. */
+  public void centerOnCurrentMovableUnit() {
+    if (lastFocusedTerritory != null) {
+      mapPanel.setUnitHighlight(
+          Collections.singleton(
+              UnitScrollerModel.getMoveableUnits(
+                  lastFocusedTerritory,
+                  movePhaseSupplier.get(),
+                  currentPlayerSupplier.get(),
+                  skippedUnits)));
+      mapPanel.centerOnTerritoryIgnoringMapLock(lastFocusedTerritory);
+    } else {
+      centerOnNextMovableUnit();
+    }
+  }
+
+  /**
+   * Skips the units in the current territory. Scroller will not scroll back to these units and the
+   * units will not be highlighted.
+   */
+  public void skipCurrentUnits() {
+    if (lastFocusedTerritory != null) {
+      skippedUnits.addAll(lastFocusedTerritory.getUnits());
+      updateMovesLeftLabel();
+    }
+    centerOnNextMovableUnit();
+  }
+
+  public void centerOnNextMovableUnit() {
+    centerOnMovableUnit(true);
+  }
+
+  public void centerOnPreviousMovableUnit() {
+    centerOnMovableUnit(false);
+  }
+
+  private void centerOnMovableUnit(final boolean selectNext) {
+    final List<Territory> allTerritories;
+    gameData.acquireReadLock();
+    try {
+      allTerritories = new ArrayList<>(gameData.getMap().getTerritories());
+    } finally {
+      gameData.releaseReadLock();
+    }
+
+    final int size = allTerritories.size();
+
+    if (!selectNext) {
+      Collections.reverse(allTerritories);
+    }
+    // new focused index is 1 greater
+    int newFocusedIndex =
+        lastFocusedTerritory == null ? 0 : allTerritories.indexOf(lastFocusedTerritory) + 1;
+    if (newFocusedIndex >= size) {
+      // if we are larger than the number of territories, we must start back at zero
+      newFocusedIndex = 0;
+    }
+    Territory newFocusedTerritory = null;
+    // make sure we go through every single territory on the board
+    int i = 0;
+    while (i < size) {
+      final Territory t = allTerritories.get(newFocusedIndex);
```

Comment:
It looks like this loop can be replaced with a foreach loop

Reply:
Sorry, I meant to comment on this section of code. It is moved from here: https://github.com/triplea-game/triplea/pull/4960/files#diff-f47d4da84dcca7433ee7489682489eb9L1673

I decided not to attempt to simplify this:
- need a unit test, without that it's somewhat reckless to make modification to 'battle-tested' code. We know it works now, if we start "fixing" it, it may no longer work. The risk of error here is pretty high, it's not trivial how it's written. With that said, adding some testing and fixing it up would be excellent and decrease maintenance costs, in this case though picking our battles I'm happy enough keeping as is and reducing risk.

Converting to a foreach loop loses the 'i' index, that would then lead to a greater refactor, I'd prefer to avoid that in favor of keeping the code intact and known working.

Reply:
The while loop construct pretty clearly can be at least a standard for loop: 32e4af2a8f38095a1f2a3bdb1995ca9ca02784be

That seems safe enough to update.